### DISCUSSION TITLE
Parse const keys to value

### DISCUSSION THREAD

**Discussion by Checksum** [colinhacks/zod#936] [Q&A] [answered]
I'm new to zod and am trying to solve a relatively simple problem. I have a const structure, which maps strings to values:

```ts
const Code = {
   'foo': 1,
   'bar': 2,
   'baz': 3
} as const;
```

I need to define the zod parser for this since the input must be validated against the enumerated keys. For example, the `code` field in  the JSON input must only be `'foo' | 'bar' | 'baz'` and I need to map it to the numeric value.

```ts
CodeParser.parse({ code: "foo" }) // pass
CodeParser.parse({ code: "something else" }) // fails
```

I tried to define a map like so, but can't really declare it as a `const` with fixed values:

```ts
const CodeEnum = z.enum(['foo', 'bar', 'baz']);
const CodeMap = z.map(CodeEnum, z.string());
```

Any pointers appreciated!

**Comment by JacobWeisenburger** [comment, marked as answer]
I think you might be thinking that [`parse`](https://github.com/colinhacks/zod#parse) takes a value and returns another value. Common mistake. But in zod, [`parse`](https://github.com/colinhacks/zod#parse) doesn't do this. You need to define a [`transform`](https://github.com/colinhacks/zod#transform) on your schema to do this. See below.

```ts
const Code = { foo: 1, bar: 2, baz: 3 }
type Key = keyof typeof Code
type Keys = [ Key, ...Key[] ]
const codeEnumSchema = z.enum( Object.keys( Code ) as Keys )
    .transform( key => Code[ key ] )
type CodeEnumInput = z.input<typeof codeEnumSchema>
// type CodeEnumInput = "foo" | "bar" | "baz"
type CodeEnumOutput = z.output<typeof codeEnumSchema>
// type CodeEnumOutput = number

codeEnumSchema.safeParse( 'foo' ) // { success: true, data: 1 }
codeEnumSchema.safeParse( 'bar' ) // { success: true, data: 2 }
codeEnumSchema.safeParse( 'baz' ) // { success: true, data: 3 }

codeEnumSchema.safeParse( 1 ) // { success: false, error: '...' }
codeEnumSchema.safeParse( 'something else' ) // { success: false, error: '...' }
```

Please let me know if I misunderstood what you are asking.
